Warehouse.Application/Features/Queries/Product/ProductList/ProductListQueryHandler.cs

```diff
@@ -0,0 +1,85 @@
+﻿using AutoMapper;
+using MediatR;
+using Microsoft.Extensions.Logging;
+using Warehouse.Application.Common.Interfaces.Persistence;
+using Warehouse.Application.Extensions;
+using Warehouse.Application.Models.Dto.ProductDtos;
+using Warehouse.Domain.Exceptions;
+using Warehouse.Persistence.EF.Extensions;
+
+namespace Warehouse.Application.Features.Queries.Product.ProductList;
+
+public record ProductListQueryHandler : IRequestHandler<ProductListGetQuery, ProductDto>
+{
+    private readonly ILogger<ProductListQueryHandler> _logger;
+    private readonly IMapper _mapper;
+    private readonly IUnitOfWork _unitOfWork;
+
+    public ProductListQueryHandler(
+        ILogger<ProductListQueryHandler> logger,
+        IMapper mapper,
+        IUnitOfWork unitOfWork)
+    {
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        _mapper = mapper ?? throw new ArgumentNullException(nameof(mapper));
+        _unitOfWork = unitOfWork;
+    }
+
+    public async Task<ProductDto> Handle(ProductListGetQuery request, CancellationToken cancellationToken)
+    {
+        var requestItems = _mapper.Map<ItemsDto>(request);
+
+        var rawProducts = await _unitOfWork.Products.GetProductsDetailsAsync();
+
+        var allProducts = _mapper.Map<IEnumerable<ProductDetailsDto>>(rawProducts);
+
+        if (allProducts is null)
```

Comment:
this check should happen before the map and also...
You have 2 ways:
1. to be in the validator, call uof and do it there.
2. At the repositories level with SingleAsync and wrap the method with try/catch and throw back your exception.